sycl/source/handler.cpp

```diff
@@ -380,23 +380,20 @@ event handler::finalize() {
     if (detail::pi::trace(detail::pi::TraceLevel::PI_TRACE_ALL)) {
       std::cout << "WARNING: An empty command group is submitted." << std::endl;
     }
-    std::shared_ptr<ext::oneapi::experimental::detail::graph_impl> GraphImpl;
     if (MGraph) {
-      GraphImpl = MGraph;
+      CommandGroup.reset(
+          new detail::CG(detail::CG::None, std::move(CGData), MCodeLoc));
+      MGraphNodeCG = std::move(CommandGroup);
     } else if (auto QueueGraph = MQueue->getCommandGraph(); QueueGraph) {
-      GraphImpl = QueueGraph;
-    }
-
-    if (GraphImpl) {
       auto EventImpl = std::make_shared<detail::event_impl>();
 
       // Extract relevant data from the handler and pass to graph to create a
       // new node representing this command group.
       std::shared_ptr<ext::oneapi::experimental::detail::node_impl> NodeImpl =
-          GraphImpl->add(CGData.MEvents);
+          QueueGraph->add(CGData.MEvents);
```

Comment:
Reviewing this code, I'm wondering if empty nodes will work with in-order queues for record & replay because we don't reach the code on lines 428-443.

Created https://github.com/reble/llvm/issues/239 to investigate this, as it shouldn't hold up this PR